### PR TITLE
HasPreviousCommit causes recursive load of commits unnecessarily

### LINKED ISSUE #13423: Request to change PR target branch (merge-base) sometimes does not work
- Gitea version (or commit ref): 1.12.5
- Git version: 2.20.1
- Operating system: Debian 10 (buster) from official binary download
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [x] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [x] No, I tried but I have no idea how to reproduce the issue.

## Description

Changing the base branch of a pull request through the UI sometimes just does nothing. I have no clue what PRs this works on and which not, so I have no reliable way of reproducing this. The POST request to `/<repo>/pull/<num>/target_branch` never produces a response. Verbose logging of everything (according to the guide from the issue template) is only this:

```
==> router.log <==
2020/11/04 14:25:02 Started POST /XXX/XXX/pull/XXX/target_branch for <IP>

==> gitea.log <==
2020/11/04 14:25:02 ...s/context/context.go:330:func1() [D] Session ID: XXX
2020/11/04 14:25:02 ...s/context/context.go:331:func1() [D] CSRF Token: XXXX
```

No response, no more logging.

Whether or not the request works depends on the PR. It never works for the same PR, but other PRs work.

### LINKED ISSUE #13684: Can't change branch in pull request
Hello!

I can't change branch in PR after PR was created.
Please take a look at this short screencast to see how it looks like: https://mttm.ml/16

Details:

- Gitea version (or commit ref): 1.13.0-rc2
- Git version: 2.20.1
- Operating system: Linux Debian 10 x64
- Database: MySQL
- Can you reproduce the bug at https://try.gitea.io: **NO** (seems to work fine there)
- Log gist (seems like no errors fro me):
```
2020/11/24 10:55:24 cmd/web.go:108:runWeb() [I] Starting Gitea on PID: 5384
2020/11/24 10:55:24 ...dules/setting/git.go:91:newGit() [I] Git Version: 2.20.1, Wire Protocol Version 2 Enabled
2020/11/24 10:55:24 routers/init.go:127:GlobalInit() [T] AppPath: /opt/gitea/gitea
2020/11/24 10:55:24 routers/init.go:128:GlobalInit() [T] AppWorkPath: /opt/gitea
2020/11/24 10:55:24 routers/init.go:129:GlobalInit() [T] Custom path: /opt/gitea/custom
2020/11/24 10:55:24 routers/init.go:130:GlobalInit() [T] Log path: /opt/gitea/log
[Macaron] 2020-11-24 10:55:33: Started GET /pulls for 80.249.94.216
[Macaron] 2020-11-24 10:55:33: Completed GET /pulls 302 Found in 782.228µs
[Macaron] 2020-11-24 10:55:33: Started GET /user/login for 80.249.94.216
[Macaron] 2020-11-24 10:55:33: Completed GET /user/login 302 Found in 7.49403ms
[Macaron] 2020-11-24 10:55:33: Started GET /pulls for 80.249.94.216
[Macaron] 2020-11-24 10:55:33: Completed GET /pulls 200 OK in 263.772952ms
[Macaron] 2020-11-24 10:55:34: Started GET /user/avatar/mishutka/-1 for 80.249.94.216
[Macaron] 2020-11-24 10:55:34: Completed GET /user/avatar/mishutka/-1 302 Found in 28.199902ms
[Macaron] 2020-11-24 10:55:34: Started GET /user/avatar/fst/-1 for 80.249.94.216
[Macaron] 2020-11-24 10:55:34: Completed GET /user/avatar/fst/-1 302 Found in 4.836308ms
[Macaron] 2020-11-24 10:55:34: Started GET /user/avatar/mitoteam/-1 for 80.249.94.216
[Macaron] 2020-11-24 10:55:34: Completed GET /user/avatar/mitoteam/-1 302 Found in 4.10068ms
[Macaron] 2020-11-24 10:55:34: Started GET /user/avatar/seniorszen/-1 for 80.249.94.216
[Macaron] 2020-11-24 10:55:34: Started GET /user/avatar/andrey/-1 for 80.249.94.216
[Macaron] 2020-11-24 10:55:34: Completed GET /user/avatar/seniorszen/-1 302 Found in 5.219339ms
[Macaron] 2020-11-24 10:55:34: Completed GET /user/avatar/andrey/-1 302 Found in 7.103264ms
[Macaron] 2020-11-24 10:55:34: Started GET /avatars/af22c6704d0d034a2a890cbd8b8107fe for 80.249.94.216
[Macaron] 2020-11-24 10:55:34: Started GET /avatars/87a45388b9b4310a5246118ccdda83b5 for 80.249.94.216
[Macaron] 2020-11-24 10:55:34: Completed GET /avatars/af22c6704d0d034a2a890cbd8b8107fe 200 OK in 1.5027ms
[Macaron] 2020-11-24 10:55:34: Completed GET /avatars/87a45388b9b4310a5246118ccdda83b5 200 OK in 2.112444ms
[Macaron] 2020-11-24 10:55:34: Started GET /user/events for 80.249.94.216
[Macaron] 2020-11-24 10:55:34: Started GET /manifest.json for 80.249.94.216
[Macaron] 2020-11-24 10:55:34: Completed GET /manifest.json 200 OK in 2.983485ms
[Macaron] 2020-11-24 10:55:34: Started GET /serviceworker.js for 80.249.94.216
[Macaron] [Static] Serving /serviceworker.js
[Macaron] 2020-11-24 10:55:34: Completed GET /serviceworker.js 304 Not Modified in 670.405µs
[Macaron] 2020-11-24 10:55:36: Started GET /serviceworker.js for 80.249.94.216
[Macaron] [Static] Serving /serviceworker.js
[Macaron] 2020-11-24 10:55:36: Completed GET /serviceworker.js 304 Not Modified in 817.972µs
[Macaron] 2020-11-24 10:55:37: Started GET /fst/tws/pulls/284 for 80.249.94.216
[Macaron] 2020-11-24 10:55:37: Completed GET /fst/tws/pulls/284 200 OK in 220.307074ms
[Macaron] 2020-11-24 10:55:37: Started GET /user/avatar/mishutka/-1 for 80.249.94.216
[Macaron] 2020-11-24 10:55:37: Started GET /user/avatar/victor/-1 for 80.249.94.216
[Macaron] 2020-11-24 10:55:37: Started GET /repo-avatars/6-9d873e9c944048a8c29650f062bf781e for 80.249.94.216
[Macaron] 2020-11-24 10:55:37: Completed GET /repo-avatars/6-9d873e9c944048a8c29650f062bf781e 200 OK in 1.604397ms
[Macaron] 2020-11-24 10:55:37: Completed GET /user/avatar/mishutka/-1 302 Found in 7.38353ms
[Macaron] 2020-11-24 10:55:37: Completed GET /user/avatar/victor/-1 302 Found in 10.252834ms
[Macaron] 2020-11-24 10:55:37: Started GET /user/avatar/andrey/-1 for 80.249.94.216
[Macaron] 2020-11-24 10:55:37: Completed GET /user/avatar/andrey/-1 302 Found in 5.244154ms
[Macaron] 2020-11-24 10:55:37: Started GET /user/avatar/bandmaster/-1 for 80.249.94.216
[Macaron] 2020-11-24 10:55:37: Completed GET /user/avatar/bandmaster/-1 302 Found in 4.893652ms
[Macaron] 2020-11-24 10:55:37: Started GET /user/avatar/doguy/-1 for 80.249.94.216
[Macaron] 2020-11-24 10:55:37: Completed GET /user/avatar/doguy/-1 302 Found in 5.393274ms
[Macaron] 2020-11-24 10:55:37: Started GET /user/avatar/office501/-1 for 80.249.94.216
[Macaron] 2020-11-24 10:55:37: Completed GET /user/avatar/office501/-1 302 Found in 6.011316ms
[Macaron] 2020-11-24 10:55:37: Started GET /user/avatar/thesis/-1 for 80.249.94.216
[Macaron] 2020-11-24 10:55:37: Started GET /user/avatar/redmine/-1 for 80.249.94.216
[Macaron] 2020-11-24 10:55:37: Started GET /avatars/35d225edcbaf7d89d91f766cbb6deb05 for 80.249.94.216
[Macaron] 2020-11-24 10:55:37: Started GET /avatars/86ee878068094f51cc14b47b54a637dc for 80.249.94.216
[Macaron] PANIC: middleware render hasn't been registered
/go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/macaron/render.go:653 (0xd8c998)
/go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/macaron/render.go:710 (0xd8c97d)
/go/src/code.gitea.io/gitea/routers/routes/routes.go:166 (0x21a7e55)
/go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/macaron/context.go:79 (0xd82a92)
/go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/inject/inject.go:157 (0xd81414)
/go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/inject/inject.go:135 (0xd811b9)
/go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/macaron/context.go:121 (0xd82c1c)
/go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/macaron/context.go:112 (0xd947a5)
/go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/macaron/recovery.go:161 (0xd94798)
/go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/macaron/logger.go:40 (0xd867d7)
/go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/inject/inject.go:157 (0xd81414)
/go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/inject/inject.go:135 (0xd811b9)
/go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/macaron/context.go:121 (0xd82c1c)
/go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/macaron/context.go:112 (0xd93aa9)
/go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/macaron/logger.go:52 (0xd93a94)
/go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/macaron/logger.go:40 (0xd867d7)
/go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/inject/inject.go:157 (0xd81414)
/go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/inject/inject.go:135 (0xd811b9)
/go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/macaron/context.go:121 (0xd82c1c)
/go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/macaron/router.go:187 (0xd959f0)
/go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/macaron/router.go:303 (0xd8f185)
/go/src/code.gitea.io/gitea/vendor/gitea.com/macaron/macaron/macaron.go:218 (0xd87b4d)
/go/src/code.gitea.io/gitea/vendor/github.com/gorilla/context/context.go:141 (0x128ea93)
/usr/local/go/src/net/http/server.go:2042 (0x777623)
/usr/local/go/src/net/http/server.go:2843 (0x77ac22)
/usr/local/go/src/net/http/server.go:1925 (0x77642c)
/usr/local/go/src/runtime/asm_amd64.s:1374 (0x478820)
[Macaron] 2020-11-24 10:55:37: Completed GET /avatars/86ee878068094f51cc14b47b54a637dc 500 Internal Server Error in 5.228459ms
[Macaron] 2020-11-24 10:55:37: Completed GET /user/avatar/thesis/-1 302 Found in 8.7415ms
[Macaron] 2020-11-24 10:55:37: Completed GET /user/avatar/redmine/-1 302 Found in 8.072278ms
[Macaron] 2020-11-24 10:55:37: Completed GET /avatars/35d225edcbaf7d89d91f766cbb6deb05 200 OK in 8.222648ms
[Macaron] 2020-11-24 10:55:37: Started GET /user/avatar/valuxin/-1 for 80.249.94.216
[Macaron] 2020-11-24 10:55:37: Completed GET /user/avatar/valuxin/-1 302 Found in 4.6403ms
[Macaron] 2020-11-24 10:55:37: Started GET /user/avatar/victor_litvin/-1 for 80.249.94.216
[Macaron] 2020-11-24 10:55:37: Completed GET /user/avatar/victor_litvin/-1 302 Found in 4.512851ms
[Macaron] 2020-11-24 10:55:38: Started GET /api/v1/repos//redmine.mito-team.com/issues/5521 for 80.249.94.216
[Macaron] 2020-11-24 10:55:38: Completed GET /api/v1/repos//redmine.mito-team.com/issues/5521 404 Not Found in 2.885575ms
[Macaron] 2020-11-24 10:55:38: Started GET /avatars/cd7557caccb37e9d67895c074f82c925 for 80.249.94.216
[Macaron] 2020-11-24 10:55:38: Completed GET /avatars/cd7557caccb37e9d67895c074f82c925 200 OK in 1.223238ms
[Macaron] 2020-11-24 10:55:38: Started GET /user/events for 80.249.94.216
[Macaron] 2020-11-24 10:55:38: Started GET /manifest.json for 80.249.94.216
[Macaron] 2020-11-24 10:55:38: Completed GET /manifest.json 200 OK in 3.100573ms
[Macaron] 2020-11-24 10:55:38: Started GET /serviceworker.js for 80.249.94.216
[Macaron] [Static] Serving /serviceworker.js
[Macaron] 2020-11-24 10:55:38: Completed GET /serviceworker.js 304 Not Modified in 767.286µs
[Macaron] 2020-11-24 10:55:39: Started GET /serviceworker.js for 80.249.94.216
[Macaron] [Static] Serving /serviceworker.js
[Macaron] 2020-11-24 10:55:39: Completed GET /serviceworker.js 304 Not Modified in 865.289µs
[Macaron] 2020-11-24 10:55:42: Started POST /fst/tws/pull/284/target_branch for 80.249.94.216
[Macaron] 2020-11-24 10:56:00: Started GET /fst/tws.git/info/refs?service=git-upload-pack for 80.249.94.216
[Macaron] 2020-11-24 10:56:00: Completed GET /fst/tws.git/info/refs?service=git-upload-pack 401 Unauthorized in 181.748213ms
[Macaron] 2020-11-24 10:56:00: Started GET /fst/tws.git/info/refs?service=git-upload-pack for 80.249.94.216
[Macaron] 2020-11-24 10:56:02: Completed GET /fst/tws.git/info/refs?service=git-upload-pack 200 OK in 1.524401452s
[Macaron] 2020-11-24 10:56:02: Started GET /mitoteam/a2m.git/info/refs?service=git-upload-pack for 46.101.160.106
[Macaron] 2020-11-24 10:56:02: Completed GET /mitoteam/a2m.git/info/refs?service=git-upload-pack 401 Unauthorized in 81.09537ms
[Macaron] 2020-11-24 10:56:02: Started GET /mitoteam/a2m.git/info/refs?service=git-upload-pack for 46.101.160.106
[Macaron] 2020-11-24 10:56:03: Completed GET /mitoteam/a2m.git/info/refs?service=git-upload-pack 200 OK in 1.513681382s
[Macaron] 2020-11-24 10:56:04: Started GET /mitoteam/checklistdynamics.git/info/refs?service=git-upload-pack for 46.101.160.106
[Macaron] 2020-11-24 10:56:04: Completed GET /mitoteam/checklistdynamics.git/info/refs?service=git-upload-pack 401 Unauthorized in 101.091536ms
[Macaron] 2020-11-24 10:56:04: Started GET /mitoteam/checklistdynamics.git/info/refs?service=git-upload-pack for 46.101.160.106
[Macaron] 2020-11-24 10:56:05: Completed GET /mitoteam/checklistdynamics.git/info/refs?service=git-upload-pack 200 OK in 1.48288532s
[Macaron] 2020-11-24 10:56:06: Started GET /mitoteam/mt-config.git/info/refs?service=git-upload-pack for 46.101.160.106
[Macaron] 2020-11-24 10:56:06: Completed GET /mitoteam/mt-config.git/info/refs?service=git-upload-pack 401 Unauthorized in 116.617878ms
[Macaron] 2020-11-24 10:56:06: Started GET /mitoteam/mt-config.git/info/refs?service=git-upload-pack for 46.101.160.106
[Macaron] 2020-11-24 10:56:07: Completed GET /mitoteam/mt-config.git/info/refs?service=git-upload-pack 404 Not Found in 1.234722716s
[Macaron] 2020-11-24 10:56:07: Started GET /fst/tws.git/info/refs?service=git-upload-pack for 46.101.160.106
[Macaron] 2020-11-24 10:56:07: Completed GET /fst/tws.git/info/refs?service=git-upload-pack 401 Unauthorized in 81.028345ms
[Macaron] 2020-11-24 10:56:07: Started GET /fst/tws.git/info/refs?service=git-upload-pack for 46.101.160.106
[Macaron] 2020-11-24 10:56:09: Completed GET /fst/tws.git/info/refs?service=git-upload-pack 200 OK in 1.533652122s
[Macaron] 2020-11-24 10:56:09: Started GET /mitoteam/velobynet.git/info/refs?service=git-upload-pack for 46.101.160.106
[Macaron] 2020-11-24 10:56:09: Completed GET /mitoteam/velobynet.git/info/refs?service=git-upload-pack 401 Unauthorized in 80.984888ms
[Macaron] 2020-11-24 10:56:09: Started GET /mitoteam/velobynet.git/info/refs?service=git-upload-pack for 46.101.160.106
[Macaron] 2020-11-24 10:56:10: Completed GET /mitoteam/velobynet.git/info/refs?service=git-upload-pack 200 OK in 1.452159587s
^C[Macaron] 2020-11-24 10:56:46: Completed GET /user/events 200 OK in 1m8.281888686s
[Macaron] 2020-11-24 10:56:46: Completed GET /user/events 200 OK in 1m12.057936251s
```

Thank you!

### PR DESCRIPTION
**Pull request by zeripath**
This PR improves HasPreviousCommit to prevent the automatic and recursive loading
of previous commits using git merge-base --is-ancestor and git rev-list

Fix #13684
Fix #13423 

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

**Comment by zeripath** [comment]
I think this is actually going to fix the associated problem.

File: modules/git/commit.go [hunk 1]
```diff
@@ -264,23 +266,26 @@ func (c *Commit) CommitsBefore() (*list.List, error) {
 
 // HasPreviousCommit returns true if a given commitHash is contained in commit's parents
 func (c *Commit) HasPreviousCommit(commitHash SHA1) (bool, error) {
-	for i := 0; i < c.ParentCount(); i++ {
-		commit, err := c.Parent(i)
+	if err := CheckGitVersionAtLeast("1.8"); err != nil {
+		_, err := NewCommand("merge-base", "--ancestor", commitHash.String(), c.ID.String()).RunInDir(c.repo.Path)
```
Comment (parnic-sks) [inline modules/git/commit.go]: Shouldn't this be `--is-ancestor`? I don't think `--ancestor` is a valid option.

**Comment by lunny** [comment]
@zeripath Could you add some tests for this?

File: modules/git/commit.go (hunk 1 above)
Comment (zeripath) [reply modules/git/commit.go]: yup it should be - the is has been deleted through fat fingers!

```suggestion
		_, err := NewCommand("merge-base", "--is-ancestor", commitHash.String(), c.ID.String()).RunInDir(c.repo.Path)
```

**Comment by zeripath** [comment]
@lunny test added

**Review by lunny — APPROVED** [review]

**Review by 6543 — APPROVED** [review]

**Comment by zeripath** [comment]
Make lgtm work

**Pull request merged**